### PR TITLE
BUG: set_levels set illegal levels.

### LINKED ISSUE #13754: BUG: failing MultiIndex.set_levels with inplace=True still modifies original index
Small example:

```
In [87]: df = pd.DataFrame(columns=pd.MultiIndex.from_product([['A','B'],['a','b']]))
```

Setting the levels with an incorrect levels (too short), fails (correctly of course):
```
In [92]: df.columns.set_levels(['c'], level=1)
...
ValueError: On level 1, label max (1) >= length of level  (1). NOTE: this index
is in an inconsistent state
```

When using `inplace=True`, this also gives the same error message, but the index *is* actually changed to this inconsistent state:

```
In [93]: df.columns.set_levels(['c'], level=1, inplace=True)
...
ValueError: On level 1, label max (1) >= length of level  (1). NOTE: this index
is in an inconsistent state

In [94]: df.columns
Out[94]:
MultiIndex(levels=[[u'A', u'B'], [u'c']],
           labels=[[0, 0, 1, 1], [0, 1, 0, 1]])
```

Just showing this `df` in the console did crash my session in the case above, in the example from #13741, it gave the error "IndexError: index 1 is out of bounds for axis 0 with size 1"

### PR DESCRIPTION
**Pull request by bkandel**
- [x] closes #13754
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

`MultiIndex.set_levels`, when given illegal level values, raises an error.
When `inplace=True`, though, the illegal level values are still accepted. This
commit fixes that behavior by checking that the proposed level values are legal
before setting them.

### CODE REVIEW

File: pandas/indexes/multi.py [hunk 1]
```diff
@@ -116,12 +116,13 @@ def __new__(cls, levels=None, labels=None, sortorder=None, names=None,
 
         return result
 
-    def _verify_integrity(self):
+    def _verify_integrity(self, new_labels=None, new_levels=None):
         """Raises ValueError if length of levels and labels don't match or any
```
Comment (jreback) [inline pandas/indexes/multi.py]: can you add a ``Parameters`` section

File: pandas/tests/indexes/test_multi.py [hunk 2]
```diff
@@ -204,6 +204,13 @@ def assert_matching(actual, expected):
         assert_matching(ind2.levels, new_levels)
         assert_matching(self.index.levels, levels)
 
+        # illegal level changing should not change levels
+        # GH 13754
+        original_index = self.index.copy()
+        with assertRaisesRegexp(ValueError, "^On"):
+            self.index.set_levels(['c'], level=0, inplace=True)
+        assert_matching(self.index.levels, original_index.levels)
+
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: can you test ``set_labels`` (with invalid labels)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/indexes/multi.py [hunk 3]
```diff
@@ -116,12 +116,23 @@ def __new__(cls, levels=None, labels=None, sortorder=None, names=None,
...
+            Labels to check for validity. Defaults to current labels.
+        new_levels : optional list
+            Levels to check for validity. Defaults to current levels.
+        """
+
+
         # NOTE: Currently does not check, among other things, that cached
         # nlevels matches nor that sortorder matches actually sortorder.
-        labels, levels = self.labels, self.levels
+        labels = new_labels or self.labels
```
Comment (jreback) [inline pandas/indexes/multi.py]: these need a check for list-like, otherwise the ``len`` check will give an odd error message. use ``is_list_like``

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/indexes/multi.py (hunk 3 above)
Comment (jreback) [inline pandas/indexes/multi.py]: also add a test which exercises this check

Comment (bkandel) [inline pandas/indexes/multi.py]: I added a test for this, but didn't add a separate check within `_verify_integrity` because the check for list-like is already present in `set_levels` and `set_labels`: https://github.com/pydata/pandas/blob/master/pandas/indexes/multi.py#L221 and https://github.com/pydata/pandas/blob/master/pandas/indexes/multi.py#L322.

**Comment by bkandel** [comment]
@jreback the Travis OSX build keeps on failing on conda install. Could you maybe delete the cache and restart the build? Or any other ideas? I'm pretty sure the failure is not caused by these changes.

File: pandas/tests/indexes/test_multi.py [hunk 4]
```diff
@@ -204,6 +204,25 @@ def assert_matching(actual, expected):
...
+        with assertRaisesRegexp(ValueError, "^On"):
+            self.index.set_labels([0, 1, 2, 3, 4, 5], level=0, inplace=True)
+        assert_matching(self.index.labels, original_index.labels)
+
+        with assertRaisesRegexp(TypeError, "^Levels"):
+            self.index.set_levels('c', level=0, inplace=True)
+        assert_matching(self.index.levels, original_index.levels)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: can you add tests for ``inplace=False`` as well (you can loop over the param I think)

**Comment by jreback** [comment]
minor comments, ping when green.

**Comment by jreback** [comment]
lgtm. ping on green.

File: pandas/tests/indexes/test_multi.py [hunk 5]
```diff
@@ -154,7 +154,7 @@ def assert_matching(actual, expected):
             # as much as possible
             self.assertEqual(len(actual), len(expected))
             for act, exp in zip(actual, expected):
-                act = np.asarray(act)
+                act = np.asarray(act, dtype=np.object_)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: why did you need to change this?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
you need to rebase as just merged #14303

File: pandas/tests/indexes/test_multi.py (hunk 5 above)
Comment (bkandel) [inline pandas/tests/indexes/test_multi.py]: Because the expected is changed to `np.object_` and if I don't change `act` also the test fails because the expected and actual are not changed.

Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: can you show me an example? it is passing now, so unclear why you had to change it.

Comment (bkandel) [inline pandas/tests/indexes/test_multi.py]: Just pushed up commit https://github.com/pydata/pandas/pull/14236/commits/e0c8897a33dc480f64cc221f71dd095aa97e0fd8 that undoes this change and fails with the error:
```
E       AssertionError: numpy array are different
E       
E       Attribute "dtype" are different
E       [left]:  int8
E       [right]: object
```

Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: int8 is right

Comment (bkandel) [inline pandas/tests/indexes/test_multi.py]: But when I take out the coercion to `np.object_`, the test `test_set_levels` fails with 
```
E       AssertionError: numpy array are different
E       
E       Attribute "dtype" are different
E       [left]:  object
E       [right]: |S4
```
This is in the test
```
        # level changing [w/o mutation]
        ind2 = self.index.set_levels(new_levels)
        assert_matching(ind2.levels, new_levels)
        assert_matching(self.index.levels, levels)
```

File: pandas/indexes/multi.py [hunk 6]
```diff
@@ -116,12 +116,22 @@ def __new__(cls, levels=None, labels=None, sortorder=None, names=None,
...
+        ----------
+        new_labels : optional list
+            Labels to check for validity. Defaults to current labels.
+        new_levels : optional list
+            Levels to check for validity. Defaults to current levels.
+        """
         # NOTE: Currently does not check, among other things, that cached
         # nlevels matches nor that sortorder matches actually sortorder.
-        labels, levels = self.labels, self.levels
+        labels = new_labels or self.labels
```
Comment (TomAugspurger) [inline pandas/indexes/multi.py]: Can `new_levels` or `new_labels` ever be a NumPy array (e.g. user passes `idx.set_levels(np.array['a', 'b'])`, or will they have been coerced to a list by now? If they're an array this will raise a ValueError. You could use `labels = new_labels if new_labels is not None else self.labels`.

Comment (TomAugspurger) [inline pandas/indexes/multi.py]: Ah, looks like it'll be a `FrozenList` by this point. Never mind then.

File: pandas/tests/indexes/test_multi.py (hunk 5 above)
Comment (bkandel) [inline pandas/tests/indexes/test_multi.py]: It looks like maybe `_ensure_index` is doing the conversion to `np.object_` in that test case (https://github.com/pydata/pandas/blob/master/pandas/indexes/base.py#L3616 and https://github.com/pydata/pandas/blob/master/pandas/lib.pyx#L1029).

Comment (bkandel) [inline pandas/tests/indexes/test_multi.py]: Fixed this issue.

**Comment by bkandel** [comment]
@jreback I believe I have addressed all of the comments.

File: doc/source/whatsnew/v0.19.0.txt [hunk 7]
```diff
@@ -1581,6 +1581,7 @@ Bug Fixes
 
 - Bug in ``eval()`` where the ``resolvers`` argument would not accept a list (:issue:`14095`)
 - Bugs in ``stack``, ``get_dummies``, ``make_axis_dummies`` which don't preserve categorical dtypes in (multi)indexes (:issue:`13854`)
-- ``PeridIndex`` can now accept ``list`` and ``array`` which contains ``pd.NaT`` (:issue:`13430`)
+- ``PeriodIndex`` can now accept ``list`` and ``array`` which contains ``pd.NaT`` (:issue:`13430`)
 - Bug in ``df.groupby`` where ``.median()`` returns arbitrary values if grouped dataframe contains empty bins (:issue:`13629`)
 - Bug in ``Index.copy()`` where ``name`` parameter was ignored (:issue:`14302`)
+- Bug in ``MultiIndex.set_levels`` where illegal level values were still set after raising an error (:issue:`13754`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.19.0.txt]: move to 0.19.1

File: pandas/indexes/multi.py [hunk 8]
```diff
@@ -116,12 +116,22 @@ def __new__(cls, levels=None, labels=None, sortorder=None, names=None,
...
+        label would exceed level bounds
+
+        Parameters
+        ----------
+        new_labels : optional list
+            Labels to check for validity. Defaults to current labels.
+        new_levels : optional list
+            Levels to check for validity. Defaults to current levels.
+        """
         # NOTE: Currently does not check, among other things, that cached
```
Comment (jreback) [inline pandas/indexes/multi.py]: can you add a ``Raises`` section

File: pandas/indexes/multi.py [hunk 9]
```diff
@@ -116,12 +116,22 @@ def __new__(cls, levels=None, labels=None, sortorder=None, names=None,
 
         return result
 
-    def _verify_integrity(self):
+    def _verify_integrity(self, new_labels=None, new_levels=None):
```
Comment (jreback) [inline pandas/indexes/multi.py]: call these ``labels, levels``. the new is confusing.

File: pandas/tests/indexes/test_multi.py [hunk 10]
```diff
@@ -149,13 +149,16 @@ def test_set_levels(self):
         levels = self.index.levels
         new_levels = [[lev + 'a' for lev in level] for level in levels]
 
-        def assert_matching(actual, expected):
+        def assert_matching(actual, expected, coerce_to_obj=True):
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: call this ``check_dtype`` and simply pass onto ``assert_numpy_array_equal`` as the same parameter name.

Comment (bkandel) [inline pandas/tests/indexes/test_multi.py]: Fixed.

File: pandas/indexes/multi.py (hunk 9 above)
Comment (bkandel) [inline pandas/indexes/multi.py]: Fixed.

File: pandas/indexes/multi.py (hunk 8 above)
Comment (bkandel) [inline pandas/indexes/multi.py]: Fixed.

File: doc/source/whatsnew/v0.19.0.txt (hunk 7 above)
Comment (bkandel) [inline doc/source/whatsnew/v0.19.0.txt]: Moved.

**Comment by bkandel** [comment]
@jreback I think all comments are addressed now.

**Comment by jreback** [comment]
thanks @bkandel lgtm.

@jorisvandenbossche can be the first back-port test!

**Review by jreback — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
@bkandel Thanks a lot!
